### PR TITLE
fix: dequeue_next_key breaks async_key_value_source TTL mechanism

### PR DESCRIPTION
**Pull request by irozzo-1A**
**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

/kind bug

> /kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

> /kind feature

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area API-version

> /area build

> /area CI

> /area driver-kmod

> /area driver-bpf

> /area driver-modern-bpf

> /area libscap-engine-bpf

> /area libscap-engine-gvisor

> /area libscap-engine-kmod

> /area libscap-engine-modern-bpf

> /area libscap-engine-nodriver

> /area libscap-engine-noop

> /area libscap-engine-source-plugin

> /area libscap-engine-savefile

> /area libscap-engine-udig

> /area libscap

> /area libpman

/area libsinsp

> /area tests

> /area proposals

**Does this PR require a change in the driver versions?**

> /version driver-API-version-major

> /version driver-API-version-minor

> /version driver-API-version-patch

> /version driver-SCHEMA-version-major

> /version driver-SCHEMA-version-minor

> /version driver-SCHEMA-version-patch

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

`async_key_value_source` provides a TTL mechanism, the problem is that when `dequeue_next_key` is invoked, there is no check for the existence of the key in `m_value_map`. If the value was pruned due to TTL expiration it gets put back by [std::map::operator[]](https://cplusplus.com/reference/map/map/operator[]/), which creates a new `lookup_request`. This may lead to an infinite loop, in case the TTL always expires.

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
fix: make sure that dequeue_next_key does not put back expired values in async_key_value_source
```

### CODE REVIEW

**Review by mstemm — DISMISSED** [review]
Sorry l missed this case. I didn’t think about successive retries and the ttl limit. Thanks for fixing it.

**Comment by irozzo-1A** [comment]
> Sorry l missed this case. I didn’t think about successive retries and the ttl limit. Thanks for fixing it.

No worries, business as usual for devs 😉 I just want to add a UT and it's ready to go

**Comment by FedeDP** [comment]
/milestone 0.11.0

We definitely want this, but i'd postpone this to the next release.

File: userspace/libsinsp/test/async_key_value_source.ut.cpp
Comment (irozzo-1A) [inline userspace/libsinsp/test/async_key_value_source.ut.cpp:75]:
```suggestion
				// Simulate failures, re-enqueue the key after m_backoff_ms milliseconds
```

File: userspace/libsinsp/container_engine/container_async_source.h
Comment (irozzo-1A) [inline userspace/libsinsp/container_engine/container_async_source.h:47]: outdated

**Comment by FedeDP** [comment]
@irozzo-1A is this PR ready? I see you left yourself some comments ;)

**Comment by irozzo-1A** [comment]
> @irozzo-1A is this PR ready? I see you left yourself some comments ;)

yeah, I have some small things to change, but mostly waiting for a review 😉

**Comment by irozzo-1A** [comment]
@FedeDP I addressed the self comments and squashed the commits, PTAL when you have some cycles 😉

**Comment by FedeDP** [comment]
This LGTM!
I want a review by people that are more knowledgeable than me on this area of the code though :) 
/cc @mstemm  @gnosek @deepskyblue86

File: userspace/libsinsp/test/async_key_value_source.ut.cpp
Comment (deepskyblue86) [inline userspace/libsinsp/test/async_key_value_source.ut.cpp:154]: maybe a [std::future](https://en.cppreference.com/w/cpp/thread/future) would be a better fit here

File: userspace/libsinsp/test/async_key_value_source.ut.cpp
Comment (deepskyblue86) [inline userspace/libsinsp/test/async_key_value_source.ut.cpp:162]: formatting is off

File: userspace/libsinsp/test/container_info.ut.cpp
Comment (deepskyblue86) [inline userspace/libsinsp/test/container_info.ut.cpp:48]:
```suggestion
	short max_retry;
	short max_delay_ms;
	std::vector<short> expected_delays;
	std::tie(max_retry, max_delay_ms, expected_delays) = GetParam();
```

**Review by deepskyblue86 — COMMENTED** [review]
@irozzo-1A thank you so much for this.
I took the liberty of suggesting some changes just because I know C++ is not your bread and butter :slightly_smiling_face:

File: userspace/libsinsp/async/async_key_value_source.tpp
Comment (irozzo-1A) [reply userspace/libsinsp/async/async_key_value_source.tpp:289]: I used `typename value_map::iterator`, for consistency but I prefer `auto` too. I will change all the occurrences in the file.

File: userspace/libsinsp/test/async_key_value_source.ut.cpp
Comment (irozzo-1A) [reply userspace/libsinsp/test/async_key_value_source.ut.cpp:154]: yep, probably I could have used [std::promise](https://en.cppreference.com/w/cpp/thread/promise), but I think this is fine too.

File: userspace/libsinsp/test/container_info.ut.cpp
Comment (irozzo-1A) [reply userspace/libsinsp/test/container_info.ut.cpp:48]: yep, I think I considered this but I don't see a substantial difference.

**Comment by irozzo-1A** [comment]
> @irozzo-1A thank you so much for this. I took the liberty of suggesting some changes just because I know C++ is not your bread and butter 🙂

Thanks for the hints 😉

**Review by FedeDP — APPROVED** [review]
/approve

**Review by Andreagit97 — APPROVED** [review]
/approve

**Pull request merged**
